crates/miden-objects/src/block/nullifier_tree.rs

```diff
@@ -2,37 +2,74 @@ use alloc::boxed::Box;
 use alloc::string::ToString;
 use alloc::vec::Vec;
 
-use miden_core::EMPTY_WORD;
 use miden_core::utils::{ByteReader, ByteWriter, Deserializable, Serializable};
+use miden_core::{EMPTY_WORD, Felt, FieldElement, Word};
 #[cfg(feature = "std")]
 use miden_crypto::merkle::{LargeSmt, LargeSmtError, SmtStorage};
 use miden_crypto::merkle::{MerkleError, MutationSet, Smt, SmtProof};
 use miden_processor::{DeserializationError, SMT_DEPTH};
 
-use crate::Word;
 use crate::block::{BlockNumber, NullifierWitness};
 use crate::errors::NullifierTreeError;
 use crate::note::Nullifier;
 
-// FREE HELPER FUNCTIONS AND CONSTANTS
+// CONSTANTS
 // ================================================================================================
 
 /// The value of an unspent nullifier in the tree.
 pub(super) const UNSPENT_NULLIFIER: Word = EMPTY_WORD;
 
-/// Returns the nullifier's leaf value in the SMT by its block number.
-pub(super) fn block_num_to_nullifier_leaf_value(block: BlockNumber) -> Word {
-    Word::from([block.as_u32(), 0, 0, 0])
+/// A nullifier block value in the [`NullifierTree`].
+///
+/// # Invariants
+///
+/// The [`NullifierBlock`] guarantees the following:
+/// - The felt at index 0 is a valid [`BlockNumber`].
+/// - The remaining felts are set to zero.
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub struct NullifierBlock(BlockNumber);
+
+impl NullifierBlock {
+    pub const UNSPENT: NullifierBlock = NullifierBlock(BlockNumber::GENESIS);
+
+    pub fn new(value: Word) -> Result<Self, NullifierTreeError> {
+        if TryInto::<u32>::try_into(value[0].as_int()).is_ok()
+            && value[1..4].iter().all(|l| l.inner() == 0)
+        {
+            return Ok(Self(BlockNumber::from(value[0].as_int() as u32)));
+        }
+        Err(NullifierTreeError::InvalidNullifierBlockNumber)
+    }
```

Comment:
I think I would also add a `TryFrom<Word> for NullifierBlock` impl that calls `new`.

For the impl, I would write it a bit simpler as:

```rust
pub fn new(word: Word) -> Result<Self, NullifierTreeError> {
    let block_num = u32::try_from(word[0].as_int())
        .map(BlockNumber::from)
        .map_err(|_| NullifierTreeError::InvalidNullifierBlockNumber)?;

    if word[1..4].iter().all(|felt| *felt == Felt::ZERO) {
        return Err(NullifierTreeError::InvalidNullifierBlockNumber);
    }

    Ok(NullifierBlock(block_num))
}
```